incubator/hnc/third_party/open-policy-agent/gatekeeper/pkg/webhook/certs.go

```diff
@@ -435,91 +452,169 @@ func (m *mapper) Map(object handler.MapObject) []reconcile.Request {
-	c, err := controller.New(("validating-webhook-controller"), mgr, controller.Options{Reconciler: r})
-	if err != nil {
-		return err
-	}
+var _ handler.Mapper = &crdMapper{}
 
-	// Watch for changes to the provided ValidatingWebhookConfiguration
-	s := &corev1.Secret{}
-	if err := c.Watch(&source.Kind{Type: s}, &handler.EnqueueRequestForObject{}); err != nil {
-		return err
+type crdMapper struct {
+	secretKey types.NamespacedName
+	crdNames  []string
+}
+
+func (m *crdMapper) Map(object handler.MapObject) []reconcile.Request {
+	if object.Meta.GetNamespace() != "" {
+		return nil
 	}
+	for _, crdName := range m.crdNames {
+		if object.Meta.GetName() == crdName {
+			return []reconcile.Request{{NamespacedName: m.secretKey}}
+		}
+	}
+	return nil
+}
 
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func addController(mgr manager.Manager, r *ReconcileWH) error {
 	vwh := &unstructured.Unstructured{}
 	vwh.SetGroupVersionKind(vwhGVK)
-	mapper := &handler.EnqueueRequestsFromMapFunc{ToRequests: &mapper{
-		secretKey: secretKey,
-		vwhKey:    vwhKey,
-	}}
-	if err := c.Watch(&source.Kind{Type: vwh}, mapper); err != nil {
+	crd := &unstructured.Unstructured{}
+	crd.SetGroupVersionKind(crdGVK)
+	// Create a new controller
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).
+		Watches(&source.Kind{Type: vwh}, &handler.EnqueueRequestsFromMapFunc{ToRequests: &mapper{
+			secretKey: r.secretKey,
+			vwhKey:    r.VWHKey,
+		}}).
+		Watches(&source.Kind{Type: crd}, &handler.EnqueueRequestsFromMapFunc{ToRequests: &crdMapper{
+			secretKey: r.secretKey,
+			crdNames:  r.CRDNames,
+		}}).Complete(r)
+	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-var _ reconcile.Reconciler = &ReconcileVWH{}
+var _ reconcile.Reconciler = &ReconcileWH{}
 
-// ReconcileVWH reconciles a validatingwebhookconfiguration, making sure it
-// has the appropriate CA cert
-type ReconcileVWH struct {
+// ReconcileWH reconciles a webhook configuration, making sure it has the
+// appropriate CA cert
+type ReconcileWH struct {
 	client    client.Client
 	scheme    *runtime.Scheme
 	ctx       context.Context
 	secretKey types.NamespacedName
-	vwhKey    types.NamespacedName
+
+	// VWHKey is the key for a ValidatingWebhookConfiguration resource.
+	VWHKey types.NamespacedName
+
+	// CRDNames is a list of CRD names with conversion webhooks.
+	CRDNames []string
 }
 
-// Reconcile reads that state of the cluster for a validatingwebhookconfiguration
-// object and makes sure the most recent CA cert is included
-func (r *ReconcileVWH) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+// Reconcile reads the state of the cluster for webhooks in validatingwebhookconfiguration
+// and CRD to make sure the most recent CA cert is included
+func (r *ReconcileWH) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	if request.NamespacedName != r.secretKey {
 		return reconcile.Result{}, nil
 	}
+
+	// Get secret and generate certs.
```

Comment:
This is a much bigger change than it used to be, right? It would be better if you could minimize the size of the change so it's very clear what's being added, since that will make it possible to upstream this code.

I think what I originally had in mind was that just underneath the call to `injectCertToWebhook` and updating the VWH, you'd just add a call to a function called something like `handlCRDWebhooks` and then _everything_ would go into that one function. That makes it much, much easier for an outside to review this diff in a few weeks.